design/notifications.md

```diff
@@ -0,0 +1,405 @@
+  listen to AMQP for incoming requests.
+- The current API will initially be implemented in the service that handles HTTP requests for backwards compatibility,
+  so that we don’t have to migrate all of our other services at once.
+- A new, more RESTful API will be implemented as well, with the goal of eventually migrating all services to the new
+  API.
+
+This change is small enough to be made without an inordinate amount of work. The service that handles HTTP requests will
+be a drop-in replacement for the existing notification agent. Sonora can be written to use the new Notification
+API. Existing DE services can use the existing API until they can be migrated at a later time.
+
+### Benefits of Switching to AMQP
+
+The first benefit of switching to AMQP is that it will eventually allow us to decouple the notification system from the
+various client services. Since the AMQP broker will be acting as an intermediary between the notification system and
+other services, it will no longer be necessary for the notification system to be running to be able to perform various
+tasks, at least not once the client services have been modified to publish events directly to AMQP. Services that are
+still using the backwards-compatible API will still have to rely on the presence of the notifications service to perform
+their tasks.
+
+The second benefit is that it allows other services to listen for these same events and perform tasks when they
+occur. Eventually, this should allow us to create a much more dynamic system that relies far less on synchronous HTTP
+requests.
+
+### Benefits of Splitting Event Recording From the REST API
+
+Aside from the usual benefits of establishing a separation of concerns, splitting the portion of the notification agent
+that records events from the portion that serves the REST API will allow us to scale the two services independently.
+
+## Component Overviews
+
+### Event Recorder
+
+This service will listen for incoming AMQP messages. It will record each message that it receives in the notifications
+database then publish related messages to two different AMQP topics: one for email requests, and one for general
+notifications. The topic for email requests will be new; email requests are currently forwarded to `iplant-email` by the
+notification agent itself. The topic for notifications already exists. This is the topic that the current DE and
+de-webhooks listen to for incoming notifications.
+
+### Email Requests
+
+This service will listen to the new AMQP topic for incoming email requests and make calls to iplant-email in order to
+actually send the messages. We may eventually want to modify other DE services to publish messages to this topic in
+order to request outgoing emails. If and when all of the other DE services are modified to use AMQP then we may want to
+modify iplant-email to listen to AMQP for incoming requests and remove this service.
+
```

Comment:
We should add support for rich email texts (links, formatting etc...)

Reply:
That's a good idea. I'll create a Jira issue for it.